src/Compilers/CSharp/Portable/Binder/LockBinder.cs

```diff
@@ -55,9 +57,91 @@ internal override BoundStatement BindLockStatementParts(BindingDiagnosticBag dia
                 hasErrors = true;
             }
 
+            if (exprType?.IsWellKnownTypeLock() == true &&
+                TryFindLockTypeInfo(exprType, diagnostics, exprSyntax) is { } lockTypeInfo)
+            {
+                // Report use-site errors for members we will use in lowering.
+                _ = diagnostics.ReportUseSite(lockTypeInfo.EnterLockScopeMethod, exprSyntax) ||
+                    diagnostics.ReportUseSite(lockTypeInfo.ScopeType, exprSyntax) ||
+                    diagnostics.ReportUseSite(lockTypeInfo.ScopeDisposeMethod, exprSyntax);
+
+                CheckRestrictedTypeInAsyncMethod(
+                    originalBinder.ContainingMemberOrLambda,
+                    lockTypeInfo.ScopeType,
+                    diagnostics,
+                    exprSyntax,
+                    errorCode: ErrorCode.ERR_BadSpecialByRefLock);
+            }
+
             BoundStatement stmt = originalBinder.BindPossibleEmbeddedStatement(_syntax.Statement, diagnostics);
             Debug.Assert(this.Locals.IsDefaultOrEmpty);
             return new BoundLockStatement(_syntax, expr, stmt, hasErrors);
         }
+
+        internal static LockTypeInfo? TryFindLockTypeInfo(TypeSymbol lockType, BindingDiagnosticBag diagnostics, SyntaxNode? syntax)
+        {
+            var enterLockScopeMethod = TryFindPublicVoidParameterlessMethod(lockType, EnterLockScopeMethodName);
+            if (!(enterLockScopeMethod is { ReturnsVoid: false, RefKind: RefKind.None }))
+            {
+                Error(diagnostics, ErrorCode.ERR_MissingPredefinedMember, syntax, LockTypeFullName, EnterLockScopeMethodName);
+                return null;
+            }
+
+            var scopeType = enterLockScopeMethod.ReturnType;
+            if (!(scopeType is { Name: LockScopeTypeName, IsValueType: true, DeclaredAccessibility: Accessibility.Public } &&
+                TypeSymbol.Equals(scopeType.ContainingType, lockType, TypeCompareKind.ConsiderEverything)))
+            {
+                Error(diagnostics, ErrorCode.ERR_MissingPredefinedMember, syntax, LockTypeFullName, EnterLockScopeMethodName);
+                return null;
+            }
+
+            var disposeMethod = TryFindPublicVoidParameterlessMethod(scopeType, WellKnownMemberNames.DisposeMethodName);
+            if (!(disposeMethod is { ReturnsVoid: true }))
+            {
+                Error(diagnostics, ErrorCode.ERR_MissingPredefinedMember, syntax, $"{LockTypeFullName}+{LockScopeTypeName}", WellKnownMemberNames.DisposeMethodName);
+                return null;
+            }
+
+            return new LockTypeInfo
```

Comment:
>return new LockTypeInfo [](http://example.com/codeflow?start=12&length=23)

It doesn't look like we are storing this instance anywhere, and simply "deconstruct" it at ```TryFindLockTypeInfo```'s call site. We could return a Nullable of a tuple, or return bool and return symbols through out parameters